lib/tls-gnutls.c

```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
+
+  for(int i = 0; ; i++) {
+    ret = gnutls_server_name_get(session, servername, &namelen, &nametype, i);
+    if (ret == GNUTLS_E_REQUESTED_DATA_NOT_AVAILABLE) {
+      /* We hit no NAME_DNS before the end, so no servername we can use. But that might be fine. */
+      free(servername);
+      servername = NULL;
+      break;
+    }
+    if (ret == GNUTLS_E_SHORT_MEMORY_BUFFER)
+      continue;  // This must not be DNS, or an invalid DNS name. Ignore
+    if (ret != GNUTLS_E_SUCCESS)
+      goto out;
+    if (nametype != GNUTLS_NAME_DNS)
+      continue;
+    /* This was a correctly retrieved DNS name. Per RFC6066, section 3, that means there
+        are no more DNS names in this request. Send this forward. */
+    break;
+  }
+
+  ret = hs->srv.cert.getcert(hs->srv.misc, servername, &cert_uri, &key_uri, &pin);
+  if (ret < 0)
+    goto out;
+
+  ret = load_certificate(session, pcert, pcert_length, pkey, cert_uri, key_uri, pin);
+
+out:
+  if (servername != NULL) {
+    free(servername);
+  }
+  if (cert_uri != NULL) {
+    explicit_bzero(cert_uri, strlen(cert_uri));
+    free(cert_uri);
+  }
+  if (key_uri != NULL) {
+    explicit_bzero(key_uri, strlen(key_uri));
+    free(key_uri);
+  }
+  if (pin != NULL) {
+    explicit_bzero(pin, strlen(pin));
+    free(pin);
+  }
+  return ret;
+}
+
+static int
+cert_clt(gnutls_session_t session, const gnutls_datum_t *req_ca_dn, int nreqs, const gnutls_pk_algorithm_t *pk_algos, int pk_algos_length, gnutls_pcert_st **pcert, unsigned int *pcert_length, gnutls_privkey_t *pkey)
+{
+  const tls_handshake_t *hs = gnutls_session_get_ptr(session);
+  char *cert_uri = NULL;
+  char *key_uri = NULL;
+  char *pin = NULL;
+  int ret = 0;
+  char **req_ca_dn_strs = gnutls_malloc(sizeof(char *) * (nreqs + 1));
+  gnutls_datum_t ca_dn;
+
+  if (req_ca_dn_strs == NULL) {
+    ret = GNUTLS_E_MEMORY_ERROR;
+    goto out;
+  }
+
+  for (int i = 0; i < nreqs; i++) {
+    ret = gnutls_x509_rdn_get2(&req_ca_dn[i], &ca_dn, 0);
+    if (ret < 0)
+      goto out;
+    req_ca_dn_strs[i] = gnutls_malloc(sizeof(char) * (ca_dn.size + 1));
+    if (req_ca_dn_strs[i] == NULL) {
+      ret = GNUTLS_E_MEMORY_ERROR;
+      goto out;
+    }
+    memcpy(req_ca_dn_strs[i], ca_dn.data, ca_dn.size);
+    req_ca_dn_strs[i][ca_dn.size] = '\0';
+  }
+  req_ca_dn_strs[nreqs+1] = NULL;
+
+  ret = hs->clt.cert.getcert(hs->clt.misc, (const char **)req_ca_dn_strs, &cert_uri, &key_uri, &pin);
+  if (ret < 0)
+    goto out;
+
+  ret = load_certificate(session, pcert, pcert_length, pkey, cert_uri, key_uri, pin);
+
+out:
+  if (ca_dn.data != NULL) {
+    explicit_bzero(ca_dn.data, ca_dn.size);
+    gnutls_free(ca_dn.data);
+  }
+  if (req_ca_dn_strs != NULL) {
+    for(int i = 0; i < nreqs; i++) {
+      if (req_ca_dn_strs[i] != NULL)
+        free(req_ca_dn_strs[i]);
+    }
+    free(req_ca_dn_strs);
+  }
+  if (cert_uri != NULL) {
+    explicit_bzero(cert_uri, strlen(cert_uri));
+    free(cert_uri);
+  }
+  if (key_uri != NULL) {
```

Comment:
This idiom of `if (thing != NULL) { explicit_bzero(thing); free(thing); }` gets repeated a lot.  Please make a wrapper for it and call that instead.